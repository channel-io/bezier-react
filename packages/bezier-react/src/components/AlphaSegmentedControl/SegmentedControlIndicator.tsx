import React, {
  useCallback,
  useState,
} from 'react'

import { useResizeDetector } from 'react-resize-detector'

import useMergeRefs from '~/src/hooks/useMergeRefs'

import * as Styled from './SegmentedControl.styled'

export function SegmentedControlIndicator({
  target,
  container,
  containerWidth,
}: {
  target: HTMLElement | null
  container: HTMLElement | null
  containerWidth?: number
}) {
  const [indicatorStyle, setIndicatorStyle] = useState<React.CSSProperties>({})

  const updateIndicatorPosition = useCallback((node: HTMLElement | null) => {
    if (node && container && target) {
      const {
        top,
        left,
        width: selectedElementWidth,
        height: selectedElementHeight,
      } = target.getBoundingClientRect()

      const {
        top: containerTop,
        left: containerLeft,
      } = container.getBoundingClientRect()

      setIndicatorStyle({
        '--bezier-react-segmented-control-indicator-transform': `translate(${left - containerLeft}px, ${top - containerTop}px)`,
        '--bezier-react-segmented-control-indicator-width': `${selectedElementWidth}px`,
        '--bezier-react-segmented-control-indicator-height': `${selectedElementHeight}px`,
      } as React.CSSProperties)
    }
    // NOTE: (@ed) to force update indicator position on container width change
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    containerWidth,
    container,
    target,
  ])

  if (!target) { return null }

  return (
    <Styled.Indicator
      ref={updateIndicatorPosition}
      style={indicatorStyle}
    />
  )
}

export function useSegmentedControlIndicator<Element extends HTMLElement>({
  target,
  refs,
}: {
  target: HTMLElement | null
  refs: React.Ref<Element>[]
}) {
  const [container, setContainer] = useState<HTMLElement | null>(null)

  const {
    width: containerWidth,
    ref: resizeDetectorRef,
  } = useResizeDetector<HTMLElement>()

  const containerRef = useMergeRefs(
    ...refs,
    resizeDetectorRef,
    setContainer as React.Ref<HTMLElement | null>,
  )

  const render = useCallback(() => (
    <SegmentedControlIndicator
      target={target}
      container={container}
      containerWidth={containerWidth}
    />
  ), [
    target,
    container,
    containerWidth,
  ])

  return {
    containerRef,
    render,
  }
}
