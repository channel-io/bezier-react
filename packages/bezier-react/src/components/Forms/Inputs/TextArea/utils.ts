import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
