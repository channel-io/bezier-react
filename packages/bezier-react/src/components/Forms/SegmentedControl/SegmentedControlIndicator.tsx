import React, {
  useCallback,
  useLayoutEffect,
  useState,
} from 'react'

import { useResizeDetector } from 'react-resize-detector'

import useMergeRefs from '~/src/hooks/useMergeRefs'

import * as Styled from './SegmentedControl.styled'

export const SEGMENTED_CONTROL_INDICATOR_TEST_ID = 'bezier-react-segmented-control-indicator'

const SegmentedControlIndicator = function SegmentedControlIndicator({
  target,
  container,
  updateKey,
}: {
  target: HTMLElement | null
  container: HTMLElement | null
  updateKey?: string
}) {
  const [indicatorNode, setIndicatorNode] = useState<HTMLDivElement | null>(null)
  const [indicatorStyle, setIndicatorStyle] = useState<React.CSSProperties>({})

  useLayoutEffect(function pauseTransitionOnMount() {
    let timer: NodeJS.Timeout | undefined

    if (indicatorNode) {
      setIndicatorStyle(prev => ({
        ...prev,
        transition: 'none',
      }))

      timer = setTimeout(() => {
        setIndicatorStyle(prev => ({
          ...prev,
          transition: undefined,
        }))
      }, Styled.indicatorTransitionMeta.duration)
    }

    return function cleanUp() {
      clearTimeout(timer)
    }
  }, [indicatorNode])

  useLayoutEffect(function updatePosition() {
    if (indicatorNode && container && target) {
      const {
        top,
        left,
        width: selectedElementWidth,
        height: selectedElementHeight,
      } = target.getBoundingClientRect()

      const {
        top: containerTop,
        left: containerLeft,
      } = container.getBoundingClientRect()

      setIndicatorStyle(prev => ({
        ...prev,
        '--bezier-react-segmented-control-indicator-transform': `translate(${left - containerLeft}px, ${top - containerTop}px)`,
        '--bezier-react-segmented-control-indicator-width': `${selectedElementWidth}px`,
        '--bezier-react-segmented-control-indicator-height': `${selectedElementHeight}px`,
      }))
    }
  }, [
    // NOTE: (@ed) to force update indicator position on container size change
    updateKey,
    indicatorNode,
    container,
    target,
  ])

  if (!target) { return null }

  return (
    <Styled.Indicator
      ref={setIndicatorNode}
      style={indicatorStyle}
      data-testid={SEGMENTED_CONTROL_INDICATOR_TEST_ID}
    />
  )
}

export function useSegmentedControlIndicator<Element extends HTMLElement>({
  target,
  refs,
}: {
  target: HTMLElement | null
  refs: React.Ref<Element>[]
}) {
  const [container, setContainer] = useState<Element | null>(null)

  const {
    width: containerWidth,
    height: containerHeight,
    ref: resizeDetectorRef,
  } = useResizeDetector<Element>()

  const containerSizeKey = `${containerWidth}-${containerHeight}`

  const containerRef = useMergeRefs(
    ...refs,
    resizeDetectorRef,
    setContainer,
  )

  const render = useCallback(() => (
    <SegmentedControlIndicator
      target={target}
      container={container}
      updateKey={containerSizeKey}
    />
  ), [
    target,
    container,
    containerSizeKey,
  ])

  return {
    containerRef,
    render,
  } as const
}
