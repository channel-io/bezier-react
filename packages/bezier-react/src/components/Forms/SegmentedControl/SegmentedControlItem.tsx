import React, {
  forwardRef,
  useCallback,
} from 'react'

import * as RadioGroupPrimitive from '@radix-ui/react-radio-group'
import * as TabsPrimitive from '@radix-ui/react-tabs'

import useMergeRefs from '~/src/hooks/useMergeRefs'
import { ariaAttr } from '~/src/utils/domUtils'

import {
  type SegmentedControlItemProps,
  type SegmentedControlProps,
  type SegmentedControlType,
} from './SegmentedControl.types'
import {
  useSegmentedControlContext,
  useSegmentedControlItemListContext,
} from './SegmentedControlContext'

import * as Styled from './SegmentedControl.styled'

/**
 * NOTE: (@ed) A property injected at runtime by the radix-ui lib.
 */
type ItemProps<Type extends SegmentedControlType> = (
  Type extends 'radiogroup'
    ? { 'data-state'?: 'unchecked' | 'checked' }
    : { 'data-state'?: 'inactive' | 'active' }
)
& React.HTMLAttributes<HTMLButtonElement>
& Partial<SegmentedControlItemProps<Type>>
& Required<Pick<SegmentedControlProps<Type, string>, 'type'>>

const Item = forwardRef<HTMLButtonElement, ItemProps<SegmentedControlType>>(function Item({
  children,
  type,
  leftContent,
  rightContent,
  ...rest
}, forwardedRef) {
  const { setSelectedElement } = useSegmentedControlItemListContext('SegmentedControlItem')

  const checked = type === 'radiogroup'
    ? (rest as ItemProps<typeof type>)?.['data-state'] === 'checked'
    : (rest as ItemProps<typeof type>)?.['data-state'] === 'active'

  const ref = useMergeRefs(
    forwardedRef,
    useCallback((node: HTMLButtonElement | null) => {
      if (checked) {
        setSelectedElement(node)
      }
    }, [
      checked,
      setSelectedElement,
    ]),
  )

  return (
    <Styled.Item
      {...rest}
      ref={ref}
      data-checked={ariaAttr(checked)}
    >
      <Styled.ItemContainer>
        { leftContent }
        <Styled.ItemLabel>
          { children }
        </Styled.ItemLabel>
        { rightContent }
      </Styled.ItemContainer>
    </Styled.Item>
  )
})

function SegmentedControlItemImpl<Value extends string>({
  children,
  ...rest
}: SegmentedControlItemProps<Value>, forwardedRef: React.Ref<HTMLButtonElement>) {
  const { type } = useSegmentedControlContext('SegmentedControlItem')

  const SegmentedControlItem = type === 'radiogroup'
    ? RadioGroupPrimitive.Item
    : TabsPrimitive.Trigger

  return (
    <SegmentedControlItem
      asChild
      ref={forwardedRef}
      {...rest}
    >
      <Item type={type}>
        { children }
      </Item>
    </SegmentedControlItem>
  )
}

/**
 * `SegmentedControlItem` component is each item in `SegmentedControl`.
 *
 * If the type of `SegmentedControl` is `radiogroup`, this component acts as a radio item.
 * In this case, it must be used as a child of `SegmentedControl`.
 *
 * If the type of `SegmentedControl` is `tabs`, this component acts as a tab item.
 * In this case, it must be used as a child of `SegmentedControlTabList`.
 */
export const SegmentedControlItem = forwardRef(SegmentedControlItemImpl) as <Value extends string>(
  props: SegmentedControlItemProps<Value> & { ref?: React.ForwardedRef<HTMLButtonElement> }
) => JSX.Element
