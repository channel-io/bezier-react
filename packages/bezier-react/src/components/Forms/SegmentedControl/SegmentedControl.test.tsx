import React from 'react'

import { isInaccessible } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { render } from '~/src/utils/test'

import {
  FormControl,
  type FormControlProps,
} from '~/src/components/Forms/FormControl'

import {
  SegmentedControl,
  SegmentedControlTabContent,
  SegmentedControlTabList,
} from './SegmentedControl'
import {
  type SegmentedControlProps,
  type SegmentedControlType,
} from './SegmentedControl.types'
import { SEGMENTED_CONTROL_INDICATOR_TEST_ID } from './SegmentedControlIndicator'
import { SegmentedControlItem } from './SegmentedControlItem'

const MOCK_UI_DATA = [
  {
    value: 'foo',
    label: 'Item 1',
    content: 'Content 1',
  },
  {
    value: 'bar',
    label: 'Item 2',
    content: 'Content 2',
  },
  {
    value: 'baz',
    label: 'Item 3',
    content: 'Content 3',
  },
]

describe('SegmentedControl >', () => {
  const renderComponent = ({
    type = 'radiogroup',
    ...rest
  }: SegmentedControlProps<SegmentedControlType, string>) => {
    const ItemList = MOCK_UI_DATA.map(({ value, label }) => (
      <SegmentedControlItem
        key={value}
        value={value}
      >
        { label }
      </SegmentedControlItem>
    ))

    return render(
      <SegmentedControl
        type={type}
        {...rest}
      >
        { type === 'radiogroup'
          ? ItemList
          : (
            <>
              <SegmentedControlTabList>
                { ItemList }
              </SegmentedControlTabList>

              { MOCK_UI_DATA.map(({ value, content }) => (
                <SegmentedControlTabContent
                  key={value}
                  value={value}
                >
                  { content }
                </SegmentedControlTabContent>
              )) }
            </>
          ) }
      </SegmentedControl>,
    )
  }

  let user: ReturnType<typeof userEvent.setup>

  beforeEach(() => {
    user = userEvent.setup()
  })

  describe('ARIA', () => {
    describe('"radiogroup" type', () => {
      it('should be accessible', () => {
        const { container } = renderComponent({ type: 'radiogroup' })
        expect(isInaccessible(container)).toBe(false)
      })

      it('should have \'role="radiogroup"\' attribute', () => {
        const { getByRole } = renderComponent({ type: 'radiogroup' })
        expect(getByRole('radiogroup')).toBeInTheDocument()
      })

      it('should be required when required prop is true', () => {
        const { getByRole } = renderComponent({ type: 'radiogroup', required: true })
        expect(getByRole('radiogroup')).toBeRequired()
      })

      it('should be disabled when disabled prop is true', () => {
        const { getByRole } = renderComponent({ type: 'radiogroup', disabled: true })
        expect(getByRole('radiogroup')).toHaveAttribute('aria-disabled', 'true')
      })

      it('children(Item) should be disabled when disabled prop is true', () => {
        const { getAllByRole } = renderComponent({ type: 'radiogroup', disabled: true })
        const items = getAllByRole('radio')
        items.forEach(item => expect(item).toBeDisabled())
      })
    })

    describe('"tabs" type', () => {
      it('should be accessible', () => {
        const { container } = renderComponent({ type: 'tabs' })
        expect(isInaccessible(container)).toBe(false)
      })

      it('should have \'role="tablist"\' attribute', () => {
        const { getByRole } = renderComponent({ type: 'tabs' })
        expect(getByRole('tablist')).toBeInTheDocument()
      })

      it('should have \'aria-orientation="horizontal"\' attribute', () => {
        const { getByRole } = renderComponent({ type: 'tabs' })
        expect(getByRole('tablist')).toHaveAttribute('aria-orientation', 'horizontal')
      })
    })
  })

  describe('Style Props', () => {
    describe('"radiogroup" type', () => {
      it('when changing the width property, the width style should be applied to the outer wrapper', () => {
        const { getByRole } = renderComponent({ type: 'radiogroup', width: '100%' })
        expect(getByRole('radiogroup')).toHaveStyle('--b-segmented-control-width: 100%')
        expect(getByRole('radiogroup')).toHaveStyle('width: var(--b-segmented-control-width)')
      })
    })

    describe('"tabs" type', () => {
      it('when changing the width property, the width style should be applied to the outer wrapper', () => {
        const { getByRole } = renderComponent({ type: 'tabs', width: '100%' })
        expect(getByRole('tablist')).toHaveStyle('--b-segmented-control-width: 100%')
        expect(getByRole('tablist')).toHaveStyle('width: var(--b-segmented-control-width)')
      })
    })
  })

  describe('User Interactions', () => {
    describe('"radiogroup" type', () => {
      it('should focus and check item when user clicks on a item', async () => {
        const { getByRole } = renderComponent({ type: 'radiogroup' })
        const item = getByRole('radio', { name: MOCK_UI_DATA[0].label })
        await user.click(item)
        expect(item).toHaveFocus()
        expect(item).toBeChecked()
        expect(item).toHaveAttribute('data-checked', 'true')
      })

      it('should call the change event handler when user clicks on a item in a controlled item group', async () => {
        const onValueChange = jest.fn()
        const { getByRole } = renderComponent({ type: 'radiogroup', onValueChange })
        const item = getByRole('radio', { name: MOCK_UI_DATA[1].label })
        await user.click(item)
        expect(onValueChange).toBeCalledTimes(1)
      })

      it('should focus on the first item item when user presses tab key', async () => {
        const { getByRole } = renderComponent({ type: 'radiogroup' })
        const item = getByRole('radio', { name: MOCK_UI_DATA[0].label })
        await user.tab()
        expect(item).toHaveFocus()
      })

      it('should check item when user presses space key on a focused item', async () => {
        const { getByRole } = renderComponent({ type: 'radiogroup' })
        const item = getByRole('radio', { name: MOCK_UI_DATA[0].label })
        await user.tab()
        expect(item).toHaveFocus()
        await user.keyboard('{ }')
        expect(item).toHaveFocus()
        expect(item).toBeChecked()
        expect(item).toHaveAttribute('data-checked', 'true')
      })

      it('should focus and check the next item item when user presses arrow down key on a item', async () => {
        const { getByRole } = renderComponent({ type: 'radiogroup' })
        const item = getByRole('radio', { name: MOCK_UI_DATA[0].label })
        const nextItem = getByRole('radio', { name: MOCK_UI_DATA[1].label })
        await user.tab()
        expect(item).toHaveFocus()
        await user.keyboard('{ArrowDown}')
        expect(nextItem).toHaveFocus()
        // FIXME: Checking behavior does not work in test environment when arrow key is pressed
        // expect(nextItem).toBeChecked()
        // expect(item).toHaveAttribute('data-checked', 'true')
      })

      it('should focus and check the next item item when user presses arrow right key on a item', async () => {
        const { getByRole } = renderComponent({ type: 'radiogroup' })
        const item = getByRole('radio', { name: MOCK_UI_DATA[0].label })
        const nextItem = getByRole('radio', { name: MOCK_UI_DATA[1].label })
        await user.tab()
        expect(item).toHaveFocus()
        await user.keyboard('{ArrowRight}')
        expect(nextItem).toHaveFocus()
        // FIXME: Checking behavior does not work in test environment when arrow key is pressed
        // expect(nextItem).toBeChecked()
        // expect(item).toHaveAttribute('data-checked', 'true')
      })

      it('should focus and check the previous item item when user presses arrow up key on a item', async () => {
        const { getByRole } = renderComponent({ type: 'radiogroup' })
        const item = getByRole('radio', { name: MOCK_UI_DATA[0].label })
        const nextItem = getByRole('radio', { name: MOCK_UI_DATA[2].label })
        await user.tab()
        expect(item).toHaveFocus()
        await user.keyboard('{ArrowUp}')
        expect(nextItem).toHaveFocus()
        // FIXME: Checking behavior does not work in test environment when arrow key is pressed
        // expect(nextItem).toBeChecked()
        // expect(item).toHaveAttribute('data-checked', 'true')
      })

      it('should focus and check the previous item item when user presses arrow left key on a item', async () => {
        const { getByRole } = renderComponent({ type: 'radiogroup' })
        const item = getByRole('radio', { name: MOCK_UI_DATA[0].label })
        const nextItem = getByRole('radio', { name: MOCK_UI_DATA[2].label })
        await user.tab()
        expect(item).toHaveFocus()
        await user.keyboard('{ArrowLeft}')
        expect(nextItem).toHaveFocus()
        // FIXME: Checking behavior does not work in test environment when arrow key is pressed
        // expect(nextItem).toBeChecked()
        // expect(item).toHaveAttribute('data-checked', 'true')
      })
    })

    describe('"tabs" type', () => {
      it('should focus and check item when user clicks on a item', async () => {
        const { getByRole } = renderComponent({ type: 'tabs' })
        const item = getByRole('tab', { name: MOCK_UI_DATA[0].label })
        await user.click(item)
        expect(item).toHaveFocus()
        expect(item).toHaveAttribute('data-checked', 'true')
      })

      it('should call the change event handler when user clicks on a item in a controlled item group', async () => {
        const onValueChange = jest.fn()
        const { getByRole } = renderComponent({ type: 'tabs', onValueChange })
        const item = getByRole('tab', { name: MOCK_UI_DATA[1].label })
        await user.click(item)
        expect(onValueChange).toBeCalledTimes(1)
      })

      it('should focus on the first item item when user presses tab key', async () => {
        const { getByRole } = renderComponent({ type: 'tabs' })
        const item = getByRole('tab', { name: MOCK_UI_DATA[0].label })
        await user.tab()
        expect(item).toHaveFocus()
        expect(item).toHaveAttribute('data-checked', 'true')
      })

      it('should check item when user presses space key on a focused item', async () => {
        const { getByRole } = renderComponent({ type: 'tabs' })
        const item = getByRole('tab', { name: MOCK_UI_DATA[0].label })
        await user.tab()
        expect(item).toHaveFocus()
        expect(item).toHaveAttribute('data-checked', 'true')
        await user.keyboard('{ }')
        expect(item).toHaveFocus()
        expect(item).toHaveAttribute('data-checked', 'true')
      })

      it('should focus and check the next item item when user presses arrow right key on a item', async () => {
        const { getByRole } = renderComponent({ type: 'tabs' })
        const item = getByRole('tab', { name: MOCK_UI_DATA[0].label })
        const nextItem = getByRole('tab', { name: MOCK_UI_DATA[1].label })
        await user.tab()
        expect(item).toHaveFocus()
        expect(item).toHaveAttribute('data-checked', 'true')
        await user.keyboard('{ArrowRight}')
        expect(nextItem).toHaveFocus()
        expect(nextItem).toHaveAttribute('data-checked', 'true')
      })

      it('should focus and check the previous item item when user presses arrow left key on a item', async () => {
        const { getByRole } = renderComponent({ type: 'tabs' })
        const item = getByRole('tab', { name: MOCK_UI_DATA[0].label })
        const nextItem = getByRole('tab', { name: MOCK_UI_DATA[2].label })
        await user.tab()
        expect(item).toHaveFocus()
        expect(item).toHaveAttribute('data-checked', 'true')
        await user.keyboard('{ArrowLeft}')
        expect(nextItem).toHaveFocus()
        expect(nextItem).toHaveAttribute('data-checked', 'true')
      })
    })
  })

  describe('Indicator', () => {
    it('has appropriate css variable in style', () => {
      const { getByTestId } = renderComponent({
        type: 'radiogroup',
        value: MOCK_UI_DATA[0].value,
      })

      const indicator = getByTestId(SEGMENTED_CONTROL_INDICATOR_TEST_ID)

      expect(indicator).toHaveStyle('width: var(--b-segmented-control-indicator-width)')
      expect(indicator).toHaveStyle('height: var(--b-segmented-control-indicator-height)')
      expect(indicator).toHaveStyle('transform: translateX(var(--b-segmented-control-indicator-translateX)) translateY(-50%)')
    })
  })

  describe('With FormControl', () => {
    describe('"radiogroup" type', () => {
      const renderComponentWithFormControl = (props: Omit<FormControlProps, 'children'>) => render(
        <FormControl {...props}>
          <SegmentedControl type="radiogroup">
            { MOCK_UI_DATA.map(({ value, label }) => (
              <SegmentedControlItem
                key={value}
                value={value}
              >
                { label }
              </SegmentedControlItem>
            )) }
          </SegmentedControl>
        </FormControl>,
      )

      it('FormControl\'s disabled prop should be passed to SegmentedControl', () => {
        const { getByRole } = renderComponentWithFormControl({ disabled: true })
        expect(getByRole('radiogroup')).toHaveAttribute('aria-disabled', 'true')
      })

      it('FormControl\'s hasError prop should be passed to SegmentedControl', () => {
        const { getByRole } = renderComponentWithFormControl({ hasError: true })
        expect(getByRole('radiogroup')).toHaveAttribute('aria-invalid', 'true')
      })

      it('FormControl\'s required prop should be passed to SegmentedControl', () => {
        const { getByRole } = renderComponentWithFormControl({ required: true })
        expect(getByRole('radiogroup')).toHaveAttribute('aria-required', 'true')
      })

      it('FormControl\'s id prop should be passed to SegmentedControl', () => {
        const { getByRole } = renderComponentWithFormControl({ id: 'form-control-id' })
        expect(getByRole('radiogroup')).toHaveAttribute('id', 'form-control-id')
      })
    })
  })
})
