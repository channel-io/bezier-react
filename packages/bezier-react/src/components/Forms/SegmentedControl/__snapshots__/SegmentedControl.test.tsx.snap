// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedControl Snapshot 1`] = `
.c4 {
  margin-top: var(--bezier-text-margin-top);
  margin-right: var(--bezier-text-margin-right);
  margin-bottom: var(--bezier-text-margin-bottom);
  margin-left: var(--bezier-text-margin-left);
  font-size: var(--bezier-text-font-size);
  font-style: var(--bezier-text-font-style);
  font-weight: var(--bezier-text-font-weight);
  line-height: var(--bezier-text-line-height);
  color: var(--bezier-text-font-color);
  -webkit-letter-spacing: var(--bezier-text-letter-spacing);
  -moz-letter-spacing: var(--bezier-text-letter-spacing);
  -ms-letter-spacing: var(--bezier-text-letter-spacing);
  letter-spacing: var(--bezier-text-letter-spacing);
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c7 {
  overflow: hidden;
  border-radius: 1px;
  background-color: #00000014;
  width: 1px;
  height: calc(100% - 12px);
  margin: 6px;
}

.c0 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 144px;
  min-width: 50px;
  overflow: hidden;
  background-color: #0000000D;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  height: 36px;
  overflow: hidden;
  border-radius: 8px;
  padding: 2px 0;
}

.c3 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: hidden;
  color: #00000066;
  cursor: pointer;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: background-color,color;
  transition-property: background-color,color;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 36px;
  overflow: hidden;
  border-radius: 6px;
}

.c3:hover {
  background-color: #0000000D;
}

.c5 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: hidden;
  color: #000000D9;
  cursor: auto;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: background-color,color;
  transition-property: background-color,color;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 36px;
  overflow: hidden;
  border-radius: 6px;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: default;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 300ms;
  transition-duration: 300ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  will-change: transform;
  height: 32px;
  overflow: hidden;
  border-radius: 6px;
}

.c2 {
  background-color: #FFFFFF;
  box-shadow: inset 0 0 2px 0 #FFFFFF1F, 0 0 2px 1px #0000000D, 0 1px 2px #00000014;
  width: 100%;
  height: 100%;
  background-color: #FFFFFF;
}

.c6 {
  position: absolute;
  top: 2px;
  bottom: 2px;
  visibility: hidden;
}

.c8 {
  position: absolute;
  top: 2px;
  bottom: 2px;
}

<div
  class="c0"
  data-testid="bezier-react-segmented-control"
>
  <div
    class="c1"
    style="width: -1.75px; transform: translateX(1.25px);"
  >
    <div
      class="c2"
    />
  </div>
  <div
    class="c3"
    data-testid="bezier-react-segmented-control-option-item-0"
    style="width: -1.75px; left: 2px;"
  >
    <span
      class="c4"
      style="--bezier-text-font-size: 1.4rem; --bezier-text-line-height: 1.8rem; --bezier-text-letter-spacing: normal; --bezier-text-font-style: normal; --bezier-text-font-weight: bold; --bezier-text-font-color: inherit; --bezier-text-margin-top: 0px; --bezier-text-margin-right: 0px; --bezier-text-margin-bottom: 0px; --bezier-text-margin-left: 0px;"
    >
      <span
        class="c4"
        style="--bezier-text-font-size: 1.5rem; --bezier-text-line-height: 2rem; --bezier-text-letter-spacing: -0.1px; --bezier-text-font-style: normal; --bezier-text-font-weight: normal; --bezier-text-font-color: inherit; --bezier-text-margin-top: 0px; --bezier-text-margin-right: 0px; --bezier-text-margin-bottom: 0px; --bezier-text-margin-left: 0px;"
      >
        0
      </span>
    </span>
  </div>
  <div
    class="c5"
    data-testid="bezier-react-segmented-control-option-item-1"
    style="width: -1.75px; left: 1.25px;"
  >
    <span
      class="c4"
      style="--bezier-text-font-size: 1.4rem; --bezier-text-line-height: 1.8rem; --bezier-text-letter-spacing: normal; --bezier-text-font-style: normal; --bezier-text-font-weight: bold; --bezier-text-font-color: inherit; --bezier-text-margin-top: 0px; --bezier-text-margin-right: 0px; --bezier-text-margin-bottom: 0px; --bezier-text-margin-left: 0px;"
    >
      <span
        class="c4"
        style="--bezier-text-font-size: 1.5rem; --bezier-text-line-height: 2rem; --bezier-text-letter-spacing: -0.1px; --bezier-text-font-style: normal; --bezier-text-font-weight: normal; --bezier-text-font-color: inherit; --bezier-text-margin-top: 0px; --bezier-text-margin-right: 0px; --bezier-text-margin-bottom: 0px; --bezier-text-margin-left: 0px;"
      >
        1
      </span>
    </span>
  </div>
  <div
    class="c3"
    data-testid="bezier-react-segmented-control-option-item-2"
    style="width: -1.75px; left: 0.5px;"
  >
    <span
      class="c4"
      style="--bezier-text-font-size: 1.4rem; --bezier-text-line-height: 1.8rem; --bezier-text-letter-spacing: normal; --bezier-text-font-style: normal; --bezier-text-font-weight: bold; --bezier-text-font-color: inherit; --bezier-text-margin-top: 0px; --bezier-text-margin-right: 0px; --bezier-text-margin-bottom: 0px; --bezier-text-margin-left: 0px;"
    >
      <span
        class="c4"
        style="--bezier-text-font-size: 1.5rem; --bezier-text-line-height: 2rem; --bezier-text-letter-spacing: -0.1px; --bezier-text-font-style: normal; --bezier-text-font-weight: normal; --bezier-text-font-color: inherit; --bezier-text-margin-top: 0px; --bezier-text-margin-right: 0px; --bezier-text-margin-bottom: 0px; --bezier-text-margin-left: 0px;"
      >
        2
      </span>
    </span>
  </div>
  <div
    class="c3"
    data-testid="bezier-react-segmented-control-option-item-3"
    style="width: -1.75px; left: -0.25px;"
  >
    <span
      class="c4"
      style="--bezier-text-font-size: 1.4rem; --bezier-text-line-height: 1.8rem; --bezier-text-letter-spacing: normal; --bezier-text-font-style: normal; --bezier-text-font-weight: bold; --bezier-text-font-color: inherit; --bezier-text-margin-top: 0px; --bezier-text-margin-right: 0px; --bezier-text-margin-bottom: 0px; --bezier-text-margin-left: 0px;"
    >
      <span
        class="c4"
        style="--bezier-text-font-size: 1.5rem; --bezier-text-line-height: 2rem; --bezier-text-letter-spacing: -0.1px; --bezier-text-font-style: normal; --bezier-text-font-weight: normal; --bezier-text-font-color: inherit; --bezier-text-margin-top: 0px; --bezier-text-margin-right: 0px; --bezier-text-margin-bottom: 0px; --bezier-text-margin-left: 0px;"
      >
        3
      </span>
    </span>
  </div>
  <div
    class="c6"
    hidden=""
    style="left: -5.75px;"
  >
    <div
      aria-orientation="vertical"
      class="c7"
      data-orientation="vertical"
      data-testid="bezier-react-divider"
      orientation="vertical"
      role="separator"
    />
  </div>
  <div
    class="c6"
    hidden=""
    style="left: -6.5px;"
  >
    <div
      aria-orientation="vertical"
      class="c7"
      data-orientation="vertical"
      data-testid="bezier-react-divider"
      orientation="vertical"
      role="separator"
    />
  </div>
  <div
    class="c8"
    style="left: -7.25px;"
  >
    <div
      aria-orientation="vertical"
      class="c7"
      data-orientation="vertical"
      data-testid="bezier-react-divider"
      orientation="vertical"
      role="separator"
    />
  </div>
</div>
`;
